sisl/io/xsf.py

```diff
@@ -384,12 +475,53 @@ class axsfSile(xsfSile):
     >>> for i in range(100):
     ...    axsf.write_geometry(geom)
     """
+    _has_steps = True
 
-    def _setup(self, *args, **kwargs):
-        # Correct number of steps
-        super()._setup(*args, **kwargs)
-        if not hasattr(self, '_md_steps'):
-            self._md_steps = 1
+    def read_geometry(self, steps=0, one_geometry=False, load_data=False):
+        """ Returns Geometries and associated data stored in the AXSF file
+
+        Parameters
+        ----------
+        steps : int or iterable of int, default 0
+            The steps to load (0-indexed). If None, load all.
+            If an integer is passed, a single Geometry is returned, and the leading dimension on data is removed.
+        one_geometry : bool, default True
+            If your axsf file contains the same geometry repeated over, you may pass
+            True here to skip construction of repeated geometries.
+        load_data : bool, optional
+            in case the AXSF file has auxiliary data, return that as well.
+
+        Returns
+        -------
+        geometries : list of Geometry or Geometry
+            A list of geometries corresponding to requested steps.
+            If one_geometry is True or steps is an integer, a single geometry (first step) is returned.
+        data : ndarray of shape (nsteps, natoms, nperatom) and dtype float64
+            Only returned if load_data is True. nsteps is the number of requested steps, not total.
+            If an integer step was passed, the leading dimension (nsteps) is removed.
+        """
+        squeeze = False
+        if isinstance(steps, int):
+            squeeze = True
+        if steps is None:
+            steps = np.arange(self.read_num_steps())
+        else:
+            steps = np.asarray(steps).ravel()
+            negs = np.flatnonzero(steps < 0)
+            steps[negs] = self.read_num_steps() + steps[negs]
+        return self._read_geom_many_step(
+            steps, reuse_geom0=one_geometry, load_data=load_data, squeeze=squeeze
+        )
+
+    @sile_fh_open()
+    def read_num_steps(self):
```

Comment:
Hmm, I don't particularly like this name... Could we perhaps make it hidden for now? Is it really useful? I.e. it seems to me the overhead of doing `read_geometry` is minimal?

Reply:
I know the name is "inelegant" but didn't know what else to do about it. I think it's nice to be able to get this information yeah. 
A possible alternative is that we read the number into `self._md_steps` upon instantiation instead?

Reply:
I would try and not do the instantiation read. That would almost always require reading the file twice.
Until we settle on a command naming convention for such a method, I think we should have it hidden. Users would then be required to read all geometries, then do `len(...)`, however, this is of minor inconvenience until the naming gets settled. E.g. the naming is useful in many other files, e.g. `outSileSiesta` and friends.

Reply:
I guess depending on what we settle on for `_md_steps` this method can be renamed to `_r_<whatever>` to keep the style of other hidden read methods.

Reply:
Yeah, good idea!